cmd/internal/cli/apptainer.go

```diff
@@ -275,27 +276,156 @@ func handleRemoteConf(remoteConfFile string) {
-		if os.IsExist(err) {
-			sylog.Debugf("%s already exists. Not creating.", confDir)
-			fi, err := os.Stat(confDir)
-			if err != nil {
-				sylog.Fatalf("Failed to retrieve information for %s: %s", confDir, err)
-			}
-			if fi.Mode().Perm() != 0o700 {
-				sylog.Debugf("Enforce permission 0700 on %s", confDir)
-				// enforce permission on user configuration directory
-				if err := os.Chmod(confDir, 0o700); err != nil {
-					// best effort as chmod could fail for various reasons (eg: readonly FS)
-					sylog.Warningf("Couldn't enforce permission 0700 on %s: %s", confDir, err)
-				}
+func handleConfDir(confDir, legacyConfigDir string) {
+	ok, err := fs.PathExists(confDir)
+	if err != nil {
+		sylog.Fatalf("Failed to retrieve information for %s: %s", confDir, err)
+	}
+
+	// apptainer user config directory exists, run perm check and return
+	if ok {
+		sylog.Debugf("%s already exists. Not creating.", confDir)
+		fi, err := os.Stat(confDir)
+		if err != nil {
+			sylog.Fatalf("Failed to retrieve information for %s: %s", confDir, err)
+		}
+		if fi.Mode().Perm() != 0o700 {
+			sylog.Debugf("Enforce permission 0700 on %s", confDir)
+			// enforce permission on user configuration directory
+			if err := os.Chmod(confDir, 0o700); err != nil {
+				// best effort as chmod could fail for various reasons (eg: readonly FS)
+				sylog.Warningf("Couldn't enforce permission 0700 on %s: %s", confDir, err)
 			}
-		} else {
-			sylog.Debugf("Could not create %s: %s", confDir, err)
 		}
-	} else {
-		sylog.Debugf("Created %s", confDir)
+		return
+	}
+
+	// apptainer user config directory doesnt exist, create it.
+	err = fs.Mkdir(confDir, 0o700)
+	if err != nil {
+		sylog.Debugf("Could not create %s: %s", confDir, err)
+		return
+	}
+
+	sylog.Debugf("Created %s", confDir)
+
+	// check if singularity user config directory exists, use it to populate configs of
+	// apptainer user config directory if it does.
+	ok, err = fs.PathExists(legacyConfigDir)
+	if err != nil {
+		sylog.Warningf("Failed to retrieve information for %s: %s", legacyConfigDir, err)
+		return
+	}
+
+	// singularity user config directory doesnt exist, return
+	if !ok {
+		return
+	}
+
+	sylog.Infof("Detected Singularity user configuration directory")
+
+	migrateRemoteConf(confDir, legacyConfigDir)
+	migrateDockerConf(confDir, legacyConfigDir)
+	migrateSypgp(confDir, legacyConfigDir)
+}
+
+func migrateRemoteConf(confDir, legacyConfigDir string) {
+	// migrate remote config
+	ok, err := fs.PathExists(syfs.LegacyRemoteConf())
+	if err != nil {
+		sylog.Warningf("Failed to retrieve information for %s: %s", syfs.LegacyRemoteConf(), err)
+		sylog.Warningf("Migration failed, you can migrate manually with \"cp -a %s %s\"", syfs.LegacyRemoteConf(), syfs.RemoteConf())
+		return
+	} else if !ok {
+		return
+	}
+
+	sylog.Infof("Detected Singularity remote configuration, migrating...")
+	err = fs.CopyFile(syfs.LegacyRemoteConf(), syfs.RemoteConf(), 0o600)
```

Comment:
I think you might want to consider checking if you can parse it into the Apptainer struct before copying.

Loading the remote config is currently not fully compatible across changes due to an UnmarshalStrict... so if a struct key gets added on the Singularity side, or removed on the Apptainer side, then it could cause issues...

https://github.com/apptainer/apptainer/blob/beb965d2e03a4b5784ac4b78af49d63fa5eb6efc/internal/pkg/remote/remote.go#L82